**Hyper parameter tuning**

- The parameters or arguments of model can change by user to get the good model performance

- Grid search

- Random search

- Assume that in Decsision tree we have following hyper parameter

    - criterion : {"gini", "entropy", "log_loss"}
        
    - splitter : {"best", "random"}
     
    - max depth : {1,2,3,4,5,6}
        
- By default we have criterion 'gini' is there

- But the model might be performed when criterion is 'entropy' also

- we need to test the combinations by providing different values

- To try the combinations we have two methods

    - Grid search:
        
        - it is a step by step by consider the all combinations
        
        - in criterion: 3,splitter: 2, max depth=6
                
        - The total combinations are : 3x2x6=36
            
    - Random search:
        
        - It will not do all 36 possibile combiantions
        
        - It will select randomly some combinations
        
- Generally we do Grid search 

- By doing this we have term called Cross validation


**Cross validation**:
    
- Cross validation: CV
    
- CV=4 means we devide data randomly into 4 parts

- Out of these 4 parts every time 3 parts consider as train data, 1 parts consider as test data

- P1  P2  P3  P4  are 4 parts of data

- Combination1:  P1 + P2 + P3 are train data  and P4 as test data
    
    - For Combination-1 will try to imporove model performance by doing Hyper parameter tuning
    
    - Assume there 36 comninations are available in DT
    
    - For the Combination1 will do all 36
 
    - so we have 36 hyper parameter comninations for each combination will have one accuracy
 
    - so we have total 36 accuracy values

- Combination2:  P1 + P2 + P4   are train data  and   P3 as test data
    
    - For Combination-2 will try to imporove model performance by doing Hyper parameter tuning
    
    - Assume there 36 comninations are available in DT
    
    - For the Combination-2 will do all 36 

- Combination3:  P1 + P3 + P4  are train data and P2 as test data
    
    - For Combination-3 will try to imporove model performance by doing Hyper parameter tuning
    
    - Assume there 36 comninations are available in DT
    
    - For the Combination-3 will do all 36 

- Combination-4: P2 + P3 + P4  are train data and P1 as test data
    
    - For Combination-4 will try to imporove model performance by doing Hyper parameter tuning
    
    - Assume there 36 comninations are available in DT
    
    - For the Combination-4 will do all 36
    
- Every combination gives 36 accurcy values corresponding each hyper parameter

- will calculate all the average accuracy of combinations individual accuracy w.r.t to each hyper parameter

- so that will get 36 average accuracy values w.r.t each hyper parameter

- Out of these 36 the maximum one is our best accuracy and corresponding hyper parameters are best parameters  

In [2]:
import pandas as pd
indexes=[f"h_{i}" for i in range(1,37)]
col1=[f"combination_1_acc_{i}" for i in range(1,37)]
col2=[f"combination_2_acc_{i}" for i in range(1,37)]
col3=[f"combination_3_acc_{i}" for i in range(1,37)]
col4=[f"combination_4_acc_{i}" for i in range(1,37)]
col5=[f"avg_acc_{i}" for i in range(1,37)]
cols=['C1','C2','C3','C4','Avg acc']
pd.DataFrame(zip(col1,col2,col3,col4,col5),index=indexes,
            columns=cols)

,C1,C2,C3,C4,Avg acc
h_1,combination_1_acc_1,combination_2_acc_1,combination_3_acc_1,combination_4_acc_1,avg_acc_1
h_2,combination_1_acc_2,combination_2_acc_2,combination_3_acc_2,combination_4_acc_2,avg_acc_2
h_3,combination_1_acc_3,combination_2_acc_3,combination_3_acc_3,combination_4_acc_3,avg_acc_3
h_4,combination_1_acc_4,combination_2_acc_4,combination_3_acc_4,combination_4_acc_4,avg_acc_4
h_5,combination_1_acc_5,combination_2_acc_5,combination_3_acc_5,combination_4_acc_5,avg_acc_5
h_6,combination_1_acc_6,combination_2_acc_6,combination_3_acc_6,combination_4_acc_6,avg_acc_6
h_7,combination_1_acc_7,combination_2_acc_7,combination_3_acc_7,combination_4_acc_7,avg_acc_7
h_8,combination_1_acc_8,combination_2_acc_8,combination_3_acc_8,combination_4_acc_8,avg_acc_8
h_9,combination_1_acc_9,combination_2_acc_9,combination_3_acc_9,combination_4_acc_9,avg_acc_9
h_10,combination_1_acc_10,combination_2_acc_10,combination_3_acc_10,combination_4_acc_10,avg_acc_10


$Step-1$

**Import the packages**

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_theme(color_codes=True)             # To get diffent different colors
pd.set_option('display.max_columns', None)  # To display the max columns
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV,cross_val_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score,\
                            precision_score,\
                            recall_score,\
                            f1_score,\
                            classification_report,\
                            roc_auc_score,roc_curve,auc
from sklearn.metrics import confusion_matrix,ConfusionMatrixDisplay

$Step-2$:
    
**Read the data**

In [6]:
data=pd.read_csv("Preprocessed_data.csv")
data

,Gender,Age,Married,Number of Dependents,Latitude,Longitude,Number of Referrals,Tenure in Months,Offer,Avg Monthly Long Distance Charges,Multiple Lines,Internet Type,Avg Monthly GB Download,Online Security,Online Backup,Device Protection Plan,Premium Tech Support,Streaming TV,Streaming Movies,Streaming Music,Unlimited Data,Contract,Paperless Billing,Payment Method,Monthly Charge,Total Charges,Total Refunds,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Customer Status
0,0,37,1,0,34.827662,-118.999073,2,9,0,42.39,0,0,16,0,1,0,1,1,0,0,1,1,1,1,65.60,593.30,0.00,0,381.51,974.81,1
1,1,46,0,0,34.162515,-118.203869,0,9,0,10.69,1,0,10,0,0,0,0,0,1,1,0,0,0,1,-4.00,542.40,38.33,10,96.21,610.28,1
2,1,50,0,0,33.645672,-117.922613,0,4,5,33.65,0,2,30,0,0,1,0,0,0,0,1,0,1,0,73.90,280.85,0.00,0,134.60,415.45,0
3,1,78,1,0,38.014457,-122.115432,1,13,4,27.82,0,2,4,0,1,1,0,1,1,0,1,0,1,0,98.00,1237.85,0.00,0,361.66,1599.51,0
4,0,75,1,0,34.227846,-119.079903,3,3,0,7.38,0,2,11,0,0,0,1,1,0,0,1,0,1,1,83.90,267.40,0.00,0,22.14,289.54,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4830,0,53,0,0,36.807595,-118.901544,0,1,5,42.09,0,2,9,0,0,0,0,0,0,0,1,0,1,1,70.15,70.15,0.00,0,42.09,112.24,0
4831,0,20,0,0,32.759327,-116.997260,0,13,4,46.68,0,1,59,1,0,0,1,0,0,1,1,1,0,1,55.15,742.90,0.00,0,606.84,1349.74,1
4832,1,40,1,0,37.734971,-120.954271,1,22,4,16.20,1,2,17,0,0,0,0,0,1,1,1,0,1,0,85.10,1873.70,0.00,0,356.40,2230.10,0
4833,1,22,0,0,39.108252,-123.645121,0,2,5,18.62,0,1,51,0,1,0,0,0,0,0,1,0,1,1,50.30,92.75,0.00,0,37.24,129.99,1


In [6]:
# data=pd.read_csv("preprocessd_data_cleaned.csv")
# data

,gender,age,no_of_days_subscribed,multi_screen,mail_subscribed,weekly_mins_watched,minimum_daily_mins,weekly_max_night_mins,videos_watched,customer_support_calls,churn
0,0,-0.264643,-0.946659,0,0,-1.508785,0.722258,-0.944404,-1.403000,-0.415473,0
1,0,0.029654,1.235009,0,0,0.304821,-0.901748,-0.688523,-0.599124,0.347522,0
2,0,2.580235,0.658246,0,0,-2.266627,0.613991,-0.483818,-1.403000,2.636507,1
3,0,-1.441834,0.783629,0,1,0.638123,-0.252146,0.079121,-0.197185,1.110517,0
4,0,0.127754,2.288228,0,0,-0.333851,0.253101,-0.893228,1.008630,-0.415473,0
...,...,...,...,...,...,...,...,...,...,...,...
3401,1,-0.805656,0.479710,0,0,0.887802,0.552184,0.865765,-0.495634,-0.415473,1
3402,0,0.477826,0.674425,0,0,2.038061,-0.931071,0.567257,2.100185,0.235227,1
3403,0,-0.550521,0.426099,0,1,-0.410669,0.344847,-0.729485,-0.403904,1.873512,1
3404,0,0.004651,-0.444637,0,0,1.359751,-0.847121,-0.715010,-0.197185,-1.178468,1


In [10]:
data.columns

Index(['Gender', 'Age', 'Married', 'Number of Dependents', 'Latitude',
       'Longitude', 'Number of Referrals', 'Tenure in Months', 'Offer',
       'Avg Monthly Long Distance Charges', 'Multiple Lines', 'Internet Type',
       'Avg Monthly GB Download', 'Online Security', 'Online Backup',
       'Device Protection Plan', 'Premium Tech Support', 'Streaming TV',
       'Streaming Movies', 'Streaming Music', 'Unlimited Data', 'Contract',
       'Paperless Billing', 'Payment Method', 'Monthly Charge',
       'Total Charges', 'Total Refunds', 'Total Extra Data Charges',
       'Total Long Distance Charges', 'Total Revenue', 'Customer Status'],
      dtype='object')

$Step-3$

**Divide data into input and output**

In [12]:
# Here the target column is Customer Status
X = data.drop('Customer Status', axis=1)
y = data['Customer Status']

$Step-4$:
    
**Train test split**

In [14]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2,random_state=0)
print(X_train.shape),
print(X_test.shape),
print(y_train.shape),
print(y_test.shape)

(3868, 30)
(967, 30)
(3868,)
(967,)


$Step-5$

**Read the base model**

In [16]:
from sklearn.model_selection import GridSearchCV,cross_val_score
from sklearn.tree import DecisionTreeClassifier
base_model_tree=DecisionTreeClassifier()   # Base model
base_model_tree

# we are not fitting the data

DecisionTreeClassifier()

$Step-6$

**Create the parameter file**

In [18]:
base_model_tree.get_params()

{'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': None,
 'max_leaf_nodes': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'monotonic_cst': None,
 'random_state': None,
 'splitter': 'best'}

**max_depth**:
    
    - The maximum depth of the tree. If None, then nodes are expanded until all leaves are pure or until all leaves contain less        than  min_samples_split samples.
    
**min_samples_split**:
     
      - The minimum number of samples required to split an internal node:
      
**min_samples_leaf**:
    
       - The minimum number of samples required to be at a leaf node.

In [20]:
# You need to create dictonary with hyper parameters
param_grid = {
    "criterion":['gini','entropy'], # 2
    'max_depth': [3, 4, 5, 6, 7, 8], # 6
    'min_samples_split': [2, 3, 4],  # 3
    'min_samples_leaf': [1, 2, 3, 4], # 4
    'random_state': [0, 42] # 2
}


# 2*6*3*4*2=288

$Step-7$

**Apply Grid Search CV**

- we have data

- we have param file

- GridSearchCV : search the each parameter

- **estimator : model name**

In [25]:
# now these parameters to apply in a DT model

grid_search = GridSearchCV(base_model_tree,  # Base model
                           param_grid, # params 
                           scoring='accuracy', # metric
                           cv=5,
                           verbose=True)

# Verbose=True I will show in Deep learning indetail
# Internal operation information will display

In [27]:
grid_search

GridSearchCV(cv=5, estimator=DecisionTreeClassifier(),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [3, 4, 5, 6, 7, 8],
                         'min_samples_leaf': [1, 2, 3, 4],
                         'min_samples_split': [2, 3, 4],
                         'random_state': [0, 42]},
             scoring='accuracy', verbose=True)

In [20]:
dir(grid_search)

['__abstractmethods__',
 '__annotations__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setstate__',
 '__sizeof__',
 '__sklearn_clone__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_abc_impl',
 '_build_request_for_signature',
 '_check_feature_names',
 '_check_n_features',
 '_check_refit_for_multimetric',
 '_doc_link_module',
 '_doc_link_template',
 '_doc_link_url_param_generator',
 '_estimator_type',
 '_format_results',
 '_get_default_requests',
 '_get_doc_link',
 '_get_metadata_request',
 '_get_param_names',
 '_get_routed_params_for_fit',
 '_get_scorers',
 '_get_tags',
 '_more_tags',
 '_parameter_constraints',
 '_repr_html_',
 '_repr_html_inner',
 '_repr_mimebundle_',
 '_required_parameters',
 '_run

$Step-8$

**Fit the data**

In [29]:
import time
start=time.time()
grid_search.fit(X_train,y_train)  
# Generally we will pass the entire data
# X_train= input data
# y_train= output data
end=time.time()
print("Total time taken is:",(end-start))

Fitting 5 folds for each of 288 candidates, totalling 1440 fits
Total time taken is: 45.27589750289917


E:\Anaconda\Lib\site-packages\numpy\ma\core.py:2820: RuntimeWarning: invalid value encountered in cast
  _data = np.array(data, dtype=dtype, copy=copy,


In [31]:
dir(grid_search)

['__abstractmethods__',
 '__annotations__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setstate__',
 '__sizeof__',
 '__sklearn_clone__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_abc_impl',
 '_build_request_for_signature',
 '_check_feature_names',
 '_check_n_features',
 '_check_refit_for_multimetric',
 '_doc_link_module',
 '_doc_link_template',
 '_doc_link_url_param_generator',
 '_estimator_type',
 '_format_results',
 '_get_default_requests',
 '_get_doc_link',
 '_get_metadata_request',
 '_get_param_names',
 '_get_routed_params_for_fit',
 '_get_scorers',
 '_get_tags',
 '_more_tags',
 '_parameter_constraints',
 '_repr_html_',
 '_repr_html_inner',
 '_repr_mimebundle_',
 '_required_parameters',
 '_run

In [24]:
grid_search.best_estimator_

DecisionTreeClassifier(criterion='entropy', max_depth=8, min_samples_leaf=4,
                       random_state=0)

In [62]:
grid_search.cv_results_

{'mean_fit_time': array([0.02429409, 0.01660342, 0.0132576 , 0.01539721, 0.01449718,
        0.01520629, 0.01303968, 0.00983634, 0.00764403, 0.00979567,
        0.00914564, 0.01165342, 0.00731025, 0.009691  , 0.00799665,
        0.00861082, 0.00815072, 0.00871034, 0.00896783, 0.00895624,
        0.00959573, 0.01120505, 0.01156316, 0.01237407, 0.01670551,
        0.01507521, 0.01441031, 0.01107917, 0.01047025, 0.01019754,
        0.01064453, 0.01113462, 0.01161146, 0.01171913, 0.01261721,
        0.0134923 , 0.01360893, 0.01360426, 0.00941424, 0.01313472,
        0.01499653, 0.0133955 , 0.01396971, 0.01349578, 0.01619787,
        0.01434898, 0.0188364 , 0.01484637, 0.02380786, 0.03062596,
        0.02799215, 0.02984347, 0.04735241, 0.02775264, 0.02385607,
        0.02769837, 0.02358971, 0.02175083, 0.02586083, 0.02312732,
        0.02421284, 0.02615929, 0.02436752, 0.02518625, 0.02624607,
        0.02668324, 0.02556014, 0.02457752, 0.02000813, 0.02080288,
        0.02197862, 0.02120209,

In [64]:
grid_search.cv_results_.keys()

dict_keys(['mean_fit_time', 'std_fit_time', 'mean_score_time', 'std_score_time', 'param_criterion', 'param_max_depth', 'param_min_samples_leaf', 'param_min_samples_split', 'param_random_state', 'params', 'split0_test_score', 'split1_test_score', 'split2_test_score', 'split3_test_score', 'split4_test_score', 'mean_test_score', 'std_test_score', 'rank_test_score'])

In [66]:
grid_search.cv_results_['params']

[{'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 1,
  'min_samples_split': 2,
  'random_state': 0},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 1,
  'min_samples_split': 2,
  'random_state': 42},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 1,
  'min_samples_split': 3,
  'random_state': 0},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 1,
  'min_samples_split': 3,
  'random_state': 42},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 1,
  'min_samples_split': 4,
  'random_state': 0},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 1,
  'min_samples_split': 4,
  'random_state': 42},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 2,
  'min_samples_split': 2,
  'random_state': 0},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 2,
  'min_samples_split': 2,
  'random_state': 42},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 2,
  'min_samples_spl

$Step-9$:
    
**Get the best parameters**

**Estimator means model**

- after developed model 1440 times 

    - 288 paramters comintaions with CV of 5

    - Each paramters combination trained 5 times

    - Ecah parameter combination will get 5 accuracy values

    - Each paramter combination will get one Avg accuracy

    - Like that we will get 288 Avg accuracy values

    - In that One paramter combination has heighest avg accurcay

    - That is best paramter

    - The model is using best parameters called as best estimator

In [68]:
grid_search.best_estimator_

# Estimator means model : 144 models

# best estimator is your model
# if you use this model 
# then we need to find all the metrics

DecisionTreeClassifier(criterion='entropy', max_depth=8, min_samples_leaf=4,
                       random_state=0)

In [70]:
grid_search.best_score_
# not a single accuracy
# avg accuracy of all 5 parts

0.8649001618996222

In [72]:
grid_search.best_params_

{'criterion': 'entropy',
 'max_depth': 8,
 'min_samples_leaf': 4,
 'min_samples_split': 2,
 'random_state': 0}

- Estimator means best Decision Tree model

- best params means the parameters which are used in best model

In [74]:
grid_search.cv_results_

{'mean_fit_time': array([0.02429409, 0.01660342, 0.0132576 , 0.01539721, 0.01449718,
        0.01520629, 0.01303968, 0.00983634, 0.00764403, 0.00979567,
        0.00914564, 0.01165342, 0.00731025, 0.009691  , 0.00799665,
        0.00861082, 0.00815072, 0.00871034, 0.00896783, 0.00895624,
        0.00959573, 0.01120505, 0.01156316, 0.01237407, 0.01670551,
        0.01507521, 0.01441031, 0.01107917, 0.01047025, 0.01019754,
        0.01064453, 0.01113462, 0.01161146, 0.01171913, 0.01261721,
        0.0134923 , 0.01360893, 0.01360426, 0.00941424, 0.01313472,
        0.01499653, 0.0133955 , 0.01396971, 0.01349578, 0.01619787,
        0.01434898, 0.0188364 , 0.01484637, 0.02380786, 0.03062596,
        0.02799215, 0.02984347, 0.04735241, 0.02775264, 0.02385607,
        0.02769837, 0.02358971, 0.02175083, 0.02586083, 0.02312732,
        0.02421284, 0.02615929, 0.02436752, 0.02518625, 0.02624607,
        0.02668324, 0.02556014, 0.02457752, 0.02000813, 0.02080288,
        0.02197862, 0.02120209,

In [34]:
grid_search.cv_results_.keys()
#grid_search.cv_results_['params']

dict_keys(['mean_fit_time', 'std_fit_time', 'mean_score_time', 'std_score_time', 'param_criterion', 'param_max_depth', 'param_min_samples_leaf', 'param_min_samples_split', 'param_random_state', 'params', 'split0_test_score', 'split1_test_score', 'split2_test_score', 'split3_test_score', 'split4_test_score', 'mean_test_score', 'std_test_score', 'rank_test_score'])

In [35]:
len(grid_search.cv_results_['param_criterion'].data)

288

In [76]:
grid_search.cv_results_['params']

[{'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 1,
  'min_samples_split': 2,
  'random_state': 0},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 1,
  'min_samples_split': 2,
  'random_state': 42},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 1,
  'min_samples_split': 3,
  'random_state': 0},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 1,
  'min_samples_split': 3,
  'random_state': 42},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 1,
  'min_samples_split': 4,
  'random_state': 0},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 1,
  'min_samples_split': 4,
  'random_state': 42},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 2,
  'min_samples_split': 2,
  'random_state': 0},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 2,
  'min_samples_split': 2,
  'random_state': 42},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 2,
  'min_samples_spl

In [80]:
grid_search.cv_results_['split0_test_score']
# p0 p1 p2 p3 = train  p5 =test
# 288 params model 288 times developed  288n accuracy

array([0.82752294, 0.82752294, 0.82752294, 0.82752294, 0.82752294,
       0.82752294, 0.82752294, 0.82752294, 0.82752294, 0.82752294,
       0.82752294, 0.82752294, 0.82752294, 0.82752294, 0.82752294,
       0.82752294, 0.82752294, 0.82752294, 0.82752294, 0.82752294,
       0.82752294, 0.82752294, 0.82752294, 0.82752294, 0.82018349,
       0.82018349, 0.82018349, 0.82018349, 0.82018349, 0.82018349,
       0.82018349, 0.82018349, 0.82018349, 0.82018349, 0.82018349,
       0.82018349, 0.82018349, 0.82018349, 0.82018349, 0.82018349,
       0.82018349, 0.82018349, 0.82201835, 0.82201835, 0.82201835,
       0.82201835, 0.82201835, 0.82201835, 0.84587156, 0.84587156,
       0.84587156, 0.84587156, 0.84587156, 0.84587156, 0.84587156,
       0.84587156, 0.84587156, 0.84587156, 0.84587156, 0.84587156,
       0.84587156, 0.84587156, 0.84587156, 0.84587156, 0.84587156,
       0.84587156, 0.84770642, 0.84770642, 0.84770642, 0.84770642,
       0.84770642, 0.84770642, 0.83853211, 0.84220183, 0.83853

In [38]:
param1_acc1=grid_search.cv_results_['split0_test_score'][0] # 288
param1_acc2=grid_search.cv_results_['split1_test_score'][0] # 288
param1_acc3=grid_search.cv_results_['split2_test_score'][0] # 288
param1_acc4=grid_search.cv_results_['split3_test_score'][0] # 288
param1_acc5=grid_search.cv_results_['split4_test_score'][0] # 288
print(param1_acc1,param1_acc2,param1_acc3,param1_acc4,param1_acc5)
param1_avg_acc=np.mean([param1_acc1,param1_acc2,param1_acc3,param1_acc4,param1_acc5])
param1_avg_acc
grid_search.cv_results_['mean_test_score'][0]==param1_avg_acc

0.8275229357798165 0.8293577981651377 0.8146788990825689 0.8256880733944955 0.8106617647058824


True

In [39]:
np.max(grid_search.cv_results_['mean_test_score'])

0.8649001618996222

In [40]:
grid_search.best_score_

0.8649001618996222

**np.argmax**

In [42]:
l1=[10,20,30,100,50]
arr1=np.array(l1)
np.max(arr1),np.argmax(arr1)

(100, 3)

In [43]:
idx=np.argmax(grid_search.cv_results_['mean_test_score'])
grid_search.cv_results_['params'][idx]

{'criterion': 'entropy',
 'max_depth': 8,
 'min_samples_leaf': 4,
 'min_samples_split': 2,
 'random_state': 0}

In [44]:
grid_search.best_params_

{'criterion': 'entropy',
 'max_depth': 8,
 'min_samples_leaf': 4,
 'min_samples_split': 2,
 'random_state': 0}

In [84]:
#len(grid_search.cv_results_['split0_test_score'])
c1=grid_search.cv_results_['split0_test_score']
c2=grid_search.cv_results_['split1_test_score']
c3=grid_search.cv_results_['split2_test_score']
c4=grid_search.cv_results_['split3_test_score']
c5=grid_search.cv_results_['split4_test_score']
mean_score=grid_search.cv_results_['mean_test_score']
df=pd.DataFrame(zip(c1,c2,c3,c4,c5,mean_score),
             index=grid_search.cv_results_['params'],
            columns=['C1','C2','C3','C4','C5','mean_test_score'])
avg=[sum(i[:-1])/5 for i in df.values]
print("The maximum avg accuracy is:",max(avg))
df['Average']=avg
df

The maximum avg accuracy is: 0.8649001618996222


,C1,C2,C3,C4,C5,mean_test_score,Average
"{'criterion': 'gini', 'max_depth': 3, 'min_samples_leaf': 1, 'min_samples_split': 2, 'random_state': 0}",0.827523,0.829358,0.814679,0.825688,0.810662,0.821582,0.821582
"{'criterion': 'gini', 'max_depth': 3, 'min_samples_leaf': 1, 'min_samples_split': 2, 'random_state': 42}",0.827523,0.829358,0.814679,0.825688,0.810662,0.821582,0.821582
"{'criterion': 'gini', 'max_depth': 3, 'min_samples_leaf': 1, 'min_samples_split': 3, 'random_state': 0}",0.827523,0.829358,0.814679,0.825688,0.810662,0.821582,0.821582
"{'criterion': 'gini', 'max_depth': 3, 'min_samples_leaf': 1, 'min_samples_split': 3, 'random_state': 42}",0.827523,0.829358,0.814679,0.825688,0.810662,0.821582,0.821582
"{'criterion': 'gini', 'max_depth': 3, 'min_samples_leaf': 1, 'min_samples_split': 4, 'random_state': 0}",0.827523,0.829358,0.814679,0.825688,0.810662,0.821582,0.821582
...,...,...,...,...,...,...,...
"{'criterion': 'entropy', 'max_depth': 8, 'min_samples_leaf': 4, 'min_samples_split': 2, 'random_state': 42}",0.877064,0.871560,0.851376,0.858716,0.845588,0.860861,0.860861
"{'criterion': 'entropy', 'max_depth': 8, 'min_samples_leaf': 4, 'min_samples_split': 3, 'random_state': 0}",0.880734,0.880734,0.849541,0.860550,0.852941,0.864900,0.864900
"{'criterion': 'entropy', 'max_depth': 8, 'min_samples_leaf': 4, 'min_samples_split': 3, 'random_state': 42}",0.877064,0.871560,0.851376,0.858716,0.845588,0.860861,0.860861
"{'criterion': 'entropy', 'max_depth': 8, 'min_samples_leaf': 4, 'min_samples_split': 4, 'random_state': 0}",0.880734,0.880734,0.849541,0.860550,0.852941,0.864900,0.864900


In [86]:
df[df['Average']==max(df['Average'])].index[0]

{'criterion': 'entropy',
 'max_depth': 8,
 'min_samples_leaf': 4,
 'min_samples_split': 2,
 'random_state': 0}

In [92]:
np.max(df['Average']), np.argmax(df['Average'])

(0.8649001618996222, 282)

In [98]:
grid_search.cv_results_['params'][282]

{'criterion': 'entropy',
 'max_depth': 8,
 'min_samples_leaf': 4,
 'min_samples_split': 2,
 'random_state': 0}

In [94]:
grid_search.best_params_

{'criterion': 'entropy',
 'max_depth': 8,
 'min_samples_leaf': 4,
 'min_samples_split': 2,
 'random_state': 0}

In [106]:
df.iloc[[282]]

,C1,C2,C3,C4,C5,mean_test_score,Average
"{'criterion': 'entropy', 'max_depth': 8, 'min_samples_leaf': 4, 'min_samples_split': 2, 'random_state': 0}",0.880734,0.880734,0.849541,0.86055,0.852941,0.8649,0.8649


In [88]:
df['Average'].max(),grid_search.best_score_

(0.8649001618996222, 0.8649001618996222)

In [48]:
df.values[0][:-1]

array([0.82752294, 0.8293578 , 0.8146789 , 0.82568807, 0.81066176,
       0.82158189])

In [49]:
# Best Parameter
df[['Average']].idxmax().values[0]

{'criterion': 'entropy',
 'max_depth': 8,
 'min_samples_leaf': 4,
 'min_samples_split': 2,
 'random_state': 0}

In [50]:
grid_search.best_params_

{'criterion': 'entropy',
 'max_depth': 8,
 'min_samples_leaf': 4,
 'min_samples_split': 2,
 'random_state': 0}

**Best Estimator means best model**

In [52]:
grid_search.best_estimator_

DecisionTreeClassifier(criterion='entropy', max_depth=8, min_samples_leaf=4,
                       random_state=0)

In [53]:
# id=df[['Average']].idxmax()
# id.values

In [54]:
# df=pd.DataFrame(zip([1,2,3],['A','B','C']),
#                columns=['Col1','Col2'])
# df['Col1'].idxmax() # index=2
# df.take([2])
# df.loc[2]

In [55]:
# # Best accuracy
# df[['Average']].max().values[0]

In [56]:
df[df['Average']==max(df['Average'])]

,C1,C2,C3,C4,C5,mean_test_score,Average
"{'criterion': 'entropy', 'max_depth': 8, 'min_samples_leaf': 4, 'min_samples_split': 2, 'random_state': 0}",0.880734,0.880734,0.849541,0.86055,0.852941,0.8649,0.8649
"{'criterion': 'entropy', 'max_depth': 8, 'min_samples_leaf': 4, 'min_samples_split': 3, 'random_state': 0}",0.880734,0.880734,0.849541,0.86055,0.852941,0.8649,0.8649
"{'criterion': 'entropy', 'max_depth': 8, 'min_samples_leaf': 4, 'min_samples_split': 4, 'random_state': 0}",0.880734,0.880734,0.849541,0.86055,0.852941,0.8649,0.8649


**Cross Validation score**

- Now we use the best model

- Perform the Cross validation score

- Best model means it takes the best parameter

- Cross validation 5 means , it gives 5 accuracy values

- Take the average of all accuracy, again that should be equal to besta accuracy

In [58]:
# Assume that we already know the best model
# best paramters with out doing Grid Search
# Now I want to apply Cross valiation
# In the cross validation every part involved in training and testing
# cross_val_score cv=5 by default

In [ ]:
cross_val_score()

In [108]:
accuracy_list=cross_val_score(grid_search.best_estimator_,
                X_train,
                y_train,
                cv=5,
                scoring='accuracy')
accuracy_list

# It will give each part accuracy
# by using best model (best estimator)

array([0.88073394, 0.88073394, 0.84954128, 0.86055046, 0.85294118])

In [112]:
accuracy_list.mean()

0.8649001618996222

**Model save**

In [117]:
import joblib
path='DT.joblib'
model=grid_search.best_estimator_
joblib.dump(model,path)

['DT.joblib']

In [61]:
## All together

#================================ Step-1: Get the base model===========================================
from sklearn.model_selection import GidSearrchCV,cross_val_score
grid_tree=DecisionTreeClassifier()   # Base model
DT_params=[keys for keys in grid_tree.get_params()]
print("avialable parameters for model:",DT_params)

print("==========================================================================================")

#================================ Step-2: Create a dictionary for params=================================
param_grid = {
    "criterion":['gini','entropy'],
    'max_depth': [3, 4, 5, 6, 7, 8],
    'min_samples_split': [2, 3, 4],
    'min_samples_leaf': [1, 2, 3, 4],
    'random_state': [0, 42]
}

print("Processing parameters are:",param_grid)

print("===============================================================================================")

#============================== Step-3: Apply the Grid serach======================================
grid_search = GridSearchCV(grid_tree,  # Base modl
                           param_grid, # params 
                           scoring='accuracy', # metric
                           cv=5,
                          verbose=True)

#============================= Step-4: fit the model with train data ===============================
grid_search.fit(X_train,y_train)

#============================= Step-5: Get the best parameters=====================================
best_params=grid_search.best_params_
best_score=grid_search.best_score_
print("best params:",best_params)
print("best train accuracy:",best_score)

ImportError: cannot import name 'GidSearrchCV' from 'sklearn.model_selection' (E:\Anaconda\Lib\site-packages\sklearn\model_selection\__init__.py)

**Identify the predictions and Metrics by using best parameters**

- Out of 1440 fittings, one best fit model is outcome

- Best fit means it will give the best parameters

- For me the best parameter combination is

    - entropy
    
    - depth=6
    
    - sample leaf=3
    
    - sampl split=2
    
    - random state=0
    
- Our goal is by using above parameter we need to find the predictions

- And will evaluate model performance

- For that Import Decision Tree and pass the best parameters

- Now fit that model with X_train, y_train

- Get the predictions by passing X_test

- Compare the predictions with y_test

- Calculate

    - Accuracy
    
    - Precision
    
    - Recall
    
    - F1 score
    
    - Roc-Auc curve

In [ ]:
grid_search.best_params_
grid_search.best_estimator_

In [ ]:
from sklearn.tree import DecisionTreeClassifier
# dtree=DecisionTreeClassifier(criterion='entropy',
#                             max_depth=8,
#                             min_samples_leaf=4,
#                             min_samples_split=2,
#                             random_state=0)
# dtree.fit(X_train, y_train)

# # =============================Step-5:  Predictions============================================================

# y_pred_dt=dtree.predict(X_test)
best_model=grid_search.best_estimator_
y_pred_dt=best_model.predict(X_test)
# ============================ Step-6: Metrics==================================================================

acc_dt= round(accuracy_score(y_test,y_pred_dt)*100,2)
f1_dt=round(f1_score(y_test,y_pred_dt),2)
precision_dt=round(precision_score(y_test,y_pred_dt),2)
recall_dt=round(recall_score(y_test,y_pred_dt),2)

print("accuray is:",acc_dt)
print("F1 is:",f1_dt)
print("Precision is:",precision_dt)
print("Recall is:",recall_dt)
print(classification_report(y_test,y_pred_dt))

# ================================Step-7:Confusion matrix=========================================================================

cmt=confusion_matrix(y_test,y_pred_dt)

disp=ConfusionMatrixDisplay(confusion_matrix=cmt,
                            display_labels = [False, True])
disp.plot()
plt.grid(False)
plt.show()


tn, fp, fn, tp = confusion_matrix(y_test,y_pred_dt).ravel()
print("True negative:",tn)
print("False postive:",fp)
print("False negative:",fn)
print("True postive:",tp)

#=======================================Step-8: ROC-AUC curve================================================================

y_dt_pred_prob=best_model.predict_proba(X_test)[:,1]   # Class-1 probabilities
fpr,tpr,threshold=roc_curve(y_test,y_dt_pred_prob) 
plt.plot([0,1],[0,1],color="navy",lw=2,label="Random-Model")
plt.plot(fpr,tpr,color="darkorange",lw=2, label="Decision-Tree Model")
plt.xlabel("False Positive Rate")
plt.ylabel("True Positive Rate")
plt.title("Receiver operating characteristic :ROC-AUC")
plt.legend()
plt.show()
print("Computed Area Under the Curve (AUC)",auc(fpr, tpr))

In [ ]:
# Decision tree will provide imprtant features also
# Information gain values 
dtree.feature_importances_

In [ ]:
imp_df = pd.DataFrame({
    "Feature Name": X_train.columns,
    "Importance": dtree.feature_importances_
})
fi = imp_df.sort_values(by="Importance", ascending=False)
fi

# This concept is valid only for DT
# you can use these features top 10 again develop the model
# Dont use these feature in another model